### PR TITLE
[types] Migrate to JSDoc

### LINKED ISSUE #499: How to have a single source of truth for types and docs?
This came out of the discussion in https://github.com/color-js/color.js/issues/496

Currently, we are generating typedoc API docs from the `.d.ts` files. However, the actual *descriptions* of API parameters live in the source `.js` files. If we point typedoc to the `.js` files we lose the types, if we point it to the TS files, we lose the descriptions. That's pretty much the definition of "stuck between a rock and hard place", both of these are TERRIBLE. 😢 

I think it should be top priority to fix this.

Solutions:

## Solution 1: We bite the bullet and move to our source to TypeScript 

Pros:
- One place for docs and types, no more having to keep separate `.d.ts` files in sync, simpler directory structure.
- Most straightforward way to solve the problem.

Cons:
- Higher bar for contributions (many color scientists can fudge together some JS, but TS can be impenetrable to someone who is not a professional front-end developer)
- No longer possible to make even the smallest edit without running build tools (currently we can get away with just running them before releases)

## Solution 2: Moving all types to JSDoc, generating `.d.ts` files from that

These days, `.d.ts` files can be [generated from JSDoc](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html), though then you're bound by what JSDoc can do, and there is a lot it cannot do, or is painful (e.g. [imported types](https://stackoverflow.com/questions/49836644/how-to-import-a-typedef-from-one-file-to-another-in-jsdoc-using-node-js)).

I don't have a good sense of what this would involve, whether it would be feasible, and how painful it might be. @jgerigmeyer @MysteryBlokHed you've worked with our types the most, what do you think?

## Solution 3: Using JSDoc’s `@inheritdoc` directive?

@MysteryBlokHed mentioned this, no idea how it can help us, perhaps they can investigate?

-----

Any other ideas?

Pinging all core maintainers @svgeesus @jgerigmeyer @MysteryBlokHed @facelessuser

### PR DESCRIPTION
**Pull request by MysteryBlokHed**
Closes #499

This PR is the same as #528, but the source branch is now on the main color.js repo to make direct modifications easier.

<hr>

This PR migrates most `.d.ts` type definitions to JSDoc type annotations instead. This is done in an effort to keep code documentation and implementation in the same place.

Here's a checklist with tasks that I can think of so far:

- [x] Files directly in `src/` converted
- [x] Files in subfolders of `src/` converted (eg. `deltaE/`)
- [x] Typing tests fixed to work with JSDoc annotations (semi-complete? runs just fine but there are type errors)

There are a few files that are currently still using `.d.ts` instead of JSDoc due to their complexity. Namely: color.d.ts, hooks.d.ts, index.d.ts, and space.d.ts.

Since the typings tests are currently not working, there's a chance that I've missed some bugs while doing this conversion. Once I find a way to get dtslint to run with JSDoc types, it'll be easier to spot any mistakes.

### CODE REVIEW

**Comment by MysteryBlokHed** [comment]
Should the release draft for v0.6.0 be updated to point to just this PR, or to point to both the original and this one? This doesn't have any of the discussion from #528, but I don't know whether we care about that for the release notes

**Comment by LeaVerou** [comment]
> Should the release draft for v0.6.0 be updated to point to just this PR, or to point to both the original and this one? This doesn't have any of the discussion from #528, if that's something we care about for the release notes

Not sure, would love more opinions. We could mention both, or we could rely on the fact that this links to #528. I’m fine either way.

File: src/ColorSpace.d.ts [hunk 1]
```diff
@@ -5,6 +5,8 @@
  */
 import { White } from "./adapt.js";
 import { ColorConstructor, Coords, ColorTypes } from "./color.js";
+import type FormatClasa from "./Format.js";
```
Comment (DmitrySharabin) [inline src/ColorSpace.d.ts:8]:
```suggestion
import type FormatClass from "./Format.js";
```

Comment (MysteryBlokHed) [reply src/ColorSpace.d.ts:8]: Whoops. Other references to this typo in the file will also need to be changed

Comment (MysteryBlokHed) [reply src/ColorSpace.d.ts:8]: Nevermind—it looks like the typo is only in the import

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+// Minimum TypeScript Version: 5.0
```
Comment (jgerigmeyer) [inline types/index.d.ts:1]: @MysteryBlokHed This fixes dtslint, which I think is now returning only real errors.

**Review by jgerigmeyer — COMMENTED** [review]
This is great! Definitely an improvement, and not surprising that it now reveals a number of existing type inconsistencies.

File: src/RGBColorSpace.js [hunk 3]
```diff
@@ -3,19 +3,17 @@ import {multiplyMatrices} from "./util.js";
...
 import XYZ_D65 from "./spaces/xyz-d65.js";
 
-/**
- * Convenience class for RGB color spaces
- * @extends {ColorSpace}
- */
+// Type "imports"
+/** @typedef {import("./types.js").RGBOptions} RGBOptions */
+
+/** Convenience class for RGB color spaces */
```
Comment (LeaVerou) [inline src/RGBColorSpace.js:9]: Why the change in formating and the removal of `@extends`?

**Review by LeaVerou — APPROVED** [review]
LGTM from a quick look. I would lave liked more locality for some of the types.js stuff, but we can explore later, it's imperative to merge this ASAP.

Thanks @DmitrySharabin for reverting the formatting changes, this was so much easier to review!

File: src/RGBColorSpace.js (hunk 3 above)
Comment (MysteryBlokHed) [reply src/RGBColorSpace.js:9]: I removed the `@extends` because TypeScript already knows due to `extends ColorSpace`. The formatting change was just because the JSDoc fit on a single line after removing `@extends`

**Comment by MysteryBlokHed** [comment]
> LGTM from a quick look. I would lave liked more locality for some of the types.js stuff, but we can explore later, it's imperative to merge this ASAP.

I put a lot of things in the `types.d.ts` because it made it a lot faster to just copy+paste from the old definitions files, but I agree that it makes sense to move definitions to their own relevant files (later).

And sorry about the formatting 😅 I've turned off that extension for this repo

**Comment by LeaVerou** [comment]
So is there any blocker remaining? Shall we merge and let it ride the trains?

**Comment by MysteryBlokHed** [comment]
> So is there any blocker remaining? Shall we merge and let it ride the trains?

I think we're all good now. `dtslint` tests are still failing, but most errors seem to be due to the tests being out of date with the API rather than actual problems that need fixing.

**Comment by MysteryBlokHed** [comment]
I or whoever else can merge as long as nobody else has any concerns

**Comment by LeaVerou** [comment]
Ship it!
